File: 2178 - 미로 탐색/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Qnode {
	int x;
	int y;
	struct Qnode* prev;
	struct Qnode* next;
} Qnode;

void enqueue(Qnode** q, int x, int y) {
	Qnode* newNode = (Qnode*)malloc(sizeof(Qnode));
	newNode->x = x;
	newNode->y = y;

	if (*q == NULL) {
		newNode->next = newNode;
		newNode->prev = newNode;
		*q = newNode;
	}
	else {
		newNode->prev = (*q)->prev;
		newNode->next = *q;
		newNode->prev->next = newNode;
		newNode->next->prev = newNode;
	}
}

void dequeue(Qnode** q) {
	Qnode* temp = *q;

	if (*q == (*q)->next) {
		*q = NULL;
	}
	else {
		(*q)->prev->next = (*q)->next;
		(*q)->next->prev = (*q)->prev;
		*q = (*q)->next;
	}

	free(temp);
}

#define MAX_N 100
#define MAX_M 100

const int dx[] = { -1, 0, 1, 0 };
const int dy[] = { 0, 1, 0, -1 };

int board[MAX_N + 2][MAX_M + 2];
int visited[MAX_N + 2][MAX_M + 2];

int N, M;
char line[MAX_M + 1];

int bfs() {
	Qnode* head = NULL;
	enqueue(&head, 1, 1);
	visited[1][1] = 1;

	while (head != NULL) {
		int x = head->x;
		int y = head->y;
		dequeue(&head);

		if (x == N && y == M) {
			return visited[x][y];
		}

		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (board[nx][ny] == 1 && visited[nx][ny] == 0) {
				enqueue(&head, nx, ny);
				visited[nx][ny] = visited[x][y] + 1;
			}
		}
	}
}

int main() {
	scanf("%d %d", &N, &M);

	for (int i = 1; i <= N; i++) {
		scanf("%s", line);

		for (int j = 1; j <= M; j++) {
			board[i][j] = line[j - 1] & 1;
		}
	}

	int result = bfs();

	printf("%d\n", result);

	return 0;
}

File: 2178 - 미로 탐색/main.py
import sys

class Qnode:
    def __init__(self, data) -> None:
        self.data = data
        self.next: 'Qnode' = None

class Queue:
    def __init__(self) -> None:
        self.__front: Qnode = None
        self.__rear: Qnode = None

    def empty(self) -> bool:
        return self.__front is None and self.__rear is None
    
    def front(self) -> any:
        if self.empty():
            return None
        else:
            return self.__front.data
    
    def enqueue(self, data) -> None:
        new_node = Qnode(data)
        if self.empty():
            self.__front = self.__rear = new_node
        else:
            self.__rear.next = new_node
            self.__rear = new_node

    def dequeue(self) -> None:
        if self.empty():
            pass
        elif self.__front is self.__rear:
            self.__front = self.__rear = None
        else:
            self.__front = self.__front.next

    def pop(self) -> any:
        result = self.front()
        self.dequeue()
        return result

DXY = [(0, 1), (1, 0), (-1, 0), (0, -1)]

board: list
n: int
m: int

def bfs():
    visited = [[0]*(m+2) for _ in range(n+2)]
    visited[1][1] = 1
    queue = Queue()
    queue.enqueue((1, 1))

    while(not queue.empty()):
        x, y = queue.pop()

        if x == n and y == m:
            return visited[x][y]
        
        for dx, dy in DXY:
            nx = x + dx
            ny = y + dy
            if board[nx][ny] == 1 and visited[nx][ny] == 0:
                queue.enqueue((nx, ny))
                visited[nx][ny] = visited[x][y] + 1

    return -1

input = sys.stdin.readline

n, m = map(int, input().strip().split())

board = [[0]*(m+2) for _ in range(n+2)]

for i in range(1, n+1):
    line = input().strip()
    for j in range(1, m+1):
        board[i][j] = int(line[j - 1])

print(bfs())
